README.md
btc-transaction
================

Create, parse, serialize Bitcoin transactions.



Install
-------

### Node.js/Browserify

    npm install --save btc-transaction


### Script

(TODO)


Usage
-----

(TODO)



Credits
-------

Most of the code from Stefan Thomas in https://github.com/bitcoinjs/bitcoinjs-lib



License
-------

(MIT License)




CHANGELOG.md
0.1.5 / 2014-02-27
------------------
* interpret coinbase prevout index as -1

0.1.3 / 2014-02-19
------------------
* When deserializing a transaction, cache the original bytes to speed up action such as getHash() and getSize(); This is because serialize() can take a very long time (sometimes in minutes) to serialize transactions with thousands of inputs

0.1.2 / 2014-02-18
-----------------
* Bugfix - serialize v2 coinbase script
* Bugfix - parsing transaction with verison 0 

0.1.1 / 2014-02-14
------------------
* Bugfix - serialize() used a removed dependency

0.1.0 / 2014-02-13
------------------
* if there is no `outpoint` default to coinbase `transactionIn`, 32 zero bytes, i.e. `0` as tx hash, `-1` as txIndex
* added `Transaction.defaultNetwork`, defaults to `mainnet` 
* commented out `analyze()` 
* commented out `getDescription()` 
* refactored `TransactionIn` and `TransactionOut` into separate files
* changed dep from `sha256` to `crypto-hashing`
* changed dep from `convert-hex` to `binstring`

0.0.3 / 2014-02-12
------------------
* Bugfix - reversing value bytes and removing bigInteger.valueOf as it does not work for numbers greater than 32 bits
* Feature- Added support to deserialize multiple transcation hex

0.0.2 / 2014-02-04
------------------
* added support for deserializing transaction hex

0.0.1 / 2014-01-13
------------------
* initial release



test/transaction.test.js
var Transaction = require('../').Transaction
var TransactionIn = require('../').TransactionIn
var TransactionOut = require('../').TransactionOut
var BigInteger = require('bigi')
var convBin = require('binstring')
var fs = require('fs')

require('terst')

describe('Transaction ', function() {
  describe(' - Transaction.deserialize()', function() {
    it(' > able to deserialize a correct buffer into a Transaction object', function() {
      // https://test.helloblock.io/v1/transactions/c68d98aaff4630ec37ca360b61a690796183e8a1b14cf123c00f0913eed6107f
      // testnet transaction
      var rawHex = '01000000010149c11ea99b6369dcd6cf9991fa0eb20a9501c7a348330e2db782e3884b9a2f000000008b483045022001c4c20a97cef3d2ff60bba1780409159e122a0b75eee99f87257a6ef3f5795e022100c5c0fd78408ebff9bc020a91e5f423994294efb18f50c2f496c4d35163acbb93014104e1934263e84e202ebffca95246b63c18c07cd369c4f02de76dbd1db89e6255dacb3ab1895af0422e24e1d1099e80f01b899cfcdf9b947575352dbc1af57466b5ffffffff01a0860100000000001976a914a5319d469e1ddd9558bd558a50e95f74b3da58c988ac00000000'
      var buf = convBin(rawHex, { in : 'hex',
        out: 'buffer'
      })
      var test = Transaction.deserialize(buf)
      T(test, 'this should return a transaction object')

      if (test) {
        T(test instanceof Transaction)
        EQ(test.version, 1)
        EQ(test.lock_time, 0)
        EQ(test.ins.length, 1)
        EQ(test.outs.length, 1)
      }

      // should be able to reserialize this into the original hex
      var hex = convBin(test.serialize(), { in : 'bytes',
        out: 'hex'
      })
      EQ(rawHex, hex)
    })

    it(' > should return false when encountering an incorrect buffer', function() {
      var rawHex = '01000000010149c11ea99b6369dcd6cf9991fa0eb20a9501c7a348330e2db7'
      var buf = convBin(rawHex, { in : 'hex',
        out: 'buffer'
      })

      var test = Transaction.deserialize(buf)
      F(test, 'this should return false')
    })

    it(' > able to parse multiple transcations', function() {
      // note: tx 1 b1fea52486ce0c62bb442b530a3f0132b826c74e473d1f2c220bfa78111c5082
      // note: tx 2 f4184fc596403b9d638783cf57adfe4c75c605f6356fbc91338530e9831e9e16

      var twoRawHex = '01000000010000000000000000000000000000000000000000000000000000000000000000ffffffff0704ffff001d0102ffffffff0100f2052a01000000434104d46c4968bde02899d2aa0963367c7a6ce34eec332b32e42e5f3407e052d64ac625da6f0718e7b302140434bd725706957c092db53805b821a85b23a7ac61725bac000000000100000001c997a5e56e104102fa209c6a852dd90660a20b2d9c352423edce25857fcd3704000000004847304402204e45e16932b8af514961a1d3a1a25fdf3f4f7732e9d624c6c61548ab5fb8cd410220181522ec8eca07de4860a4acdd12909d831cc56cbbac4622082221a8768d1d0901ffffffff0200ca9a3b00000000434104ae1a62fe09c5f51b13905f07f06b99a2f7159b2225f374cd378d71302fa28414e7aab37397f554a7df5f142c21c1b7303b8a0626f1baded5c72a704f7e6cd84cac00286bee0000000043410411db93e1dcdb8a016b49840f8c53bc1eb68a382e97b1482ecad7b148a6909a5cb2e0eaddfb84ccf9744464f82e160bfa9b8b64f9d4c03f999b8643f656b412a3ac00000000'

      var buf = convBin(twoRawHex, { in : 'hex',
        out: 'buffer'
      })
      var test = Transaction.deserialize(buf, 2)
      T(test, 'this should not return false')
      T(Array.isArray(test), 'this should return an Array of transaction objects')

      if (test) {
        EQ(test.length, 2)
        T(test[0] instanceof Transaction)
        T(test[1] instanceof Transaction)
        EQ(test[0].version, 1)
        EQ(test[1].version, 1)
        EQ(test[0].lock_time, 0)
        EQ(test[1].lock_time, 0)
        EQ(test[0].ins.length, 1)
        EQ(test[1].ins.length, 1)
        EQ(test[0].outs.length, 1)
        EQ(test[1].outs.length, 2)
      }
    })

    it(' > supports BIP34 coinbase transactions', function() {
      var coinbaseHex = '01000000010000000000000000000000000000000000000000000000000000000000000000ffffffff0d03d891000450a8eb0b4ed20100ffffffff0100f2052a010000001976a9148c91773ffc9541c8f871ba62d3e7df2c54eddd7488ac00000000'
      var buf = convBin(coinbaseHex, { in : 'hex',
        out: 'buffer'
      })
      var v2coinbaseTx = Transaction.deserialize(buf)
      T(v2coinbaseTx)
      EQ(v2coinbaseTx.ins[0].script.getBlockHeight(), 37336)
      EQ(v2coinbaseTx.ins[0].outpoint.index, -1)
    })

    it(' > parse transaction with version of 0', function() {
      // https://helloblock.io/testnet/transactions/f1ebaf4604158a8a385cb55d1c0ca5efde98ab2d6f92bc667a02485f258c064e
      var hex = '000000000100eb1d909b9f297fa7f72dc7d7298f19d5454c1cae28d7463f150a60aa2814d9010000006a47304402207413f23d980d48ba4855126ab8fc896fdeeb7752cd9c76a2054a696a5a6137cc02201a674d6cfd32668b9ac4c50ca853975a085f724cb92e95a189610e686e7f3204012103ac81c3203de55b31478da413d9bb68b99dc8e33176f9f48e5efcc0900bb41b4affffffff024e61bc00000000001976a914cdf39308e3b7ad69de09e1e4d99e036905a4289688aca2583905000000001976a914a2e4f051244a24e469d28f076fec1db4f79512b788ac00000000'
      var buf = convBin(hex, { in : 'hex',
        out: 'buffer'
      })
      var test = Transaction.deserialize(buf, 1)
      EQ(convBin(test.serialize(), { in : 'bytes',
        out: 'hex'
      }), hex)
    })

    it(' > caches the original bytes so that we dont need to serialize again to getHash() or getSize() ', function() {
      var fileBuf = fs.readFileSync('./test/fixtures/95ea61f319ed0d2b28e94cb0164396b4024bc6ad624fcb492c5c87a088592e81.bin')

      var tx = Transaction.deserialize(fileBuf)
      // note: serializing a large transaction (1000+ inputs) such as this can take up to 2 minutes. So we have to cache the original bytes when we deserialize it so that getHash() will be fast
      EQ(convBin(tx.getHash(), { in : 'bytes',
        out: 'hex'
      }), '95ea61f319ed0d2b28e94cb0164396b4024bc6ad624fcb492c5c87a088592e81')
      EQ(tx.getSize(), 389406)
    })

  })

  describe(' - Transaction.getTotalOutValue()', function() {
    it(' > should return the correct value (from LE byte array)', function() {
      Transaction.defaultNetwork = 'testnet'
      var test = new Transaction()
      test.addOutput('mvaRDyLUeF4CP7Lu9umbU3FxehyC5nUz3L', '10000000000')
      test.addOutput('mvaRDyLUeF4CP7Lu9umbU3FxehyC5nUz3L', '20000000000')

      T(test.getTotalOutValue().compareTo(BigInteger('30000000000', 10)) === 0)
    })
  })
})

describe('TransactionIn', function() {
  describe(' - new TransactionIn()', function() {
    it(' > Should be able to generate a coinbase TranscationIn', function() {
      Transaction.defaultNetwork = 'mainnet'
      // Genesis Coinbase Transaction
      // https://helloblock.io/mainnet/transactions/0e3e2357e806b6cdb1f70b54c3a3a17b6714ee1f0e68bebb44a74b1efd512098

      // scenario 1:
      // there is outpoint but there is no hash
      var coinbaseInput1 = new TransactionIn({
        outpoint: {
          hash: null,
          index: null
        },
        script: '04ffff001d0104'
      })

      EQ(coinbaseInput1.outpoint.hash, '0000000000000000000000000000000000000000000000000000000000000000')
      EQ(coinbaseInput1.outpoint.index, -1)

      // scenario 2:
      // there no outpoint
      var coinbaseInput2 = new TransactionIn({
        script: '04ffff001d0104'
      })

      EQ(coinbaseInput2.outpoint.hash, '0000000000000000000000000000000000000000000000000000000000000000')
      EQ(coinbaseInput2.outpoint.index, -1)
    })
  })
})

describe('TransactionOut', function() {
  describe(' - new TransactionOut()', function() {
    it(' > Should generate the consistent value bytes order (should be unsigned byte array padded with zero and reversed)', function() {
      Transaction.defaultNetwork = 'testnet'
      // https://test.helloblock.io/v1/transactions/c68d98aaff4630ec37ca360b61a690796183e8a1b14cf123c00f0913eed6107f
      // first output
      // {
      // n: 0,
      // value: 100000,
      // scriptPubKey: "76a914a5319d469e1ddd9558bd558a50e95f74b3da58c988ac",
      // spent: false,
      // address: "mvaRDyLUeF4CP7Lu9umbU3FxehyC5nUz3L",
      // }

      var testOutput = new TransactionOut({
        script: '76a914a5319d469e1ddd9558bd558a50e95f74b3da58c988ac',
        value: '100000'
      })

      T(testOutput.value.length == 8)

      // should be unsigned byte padded with zero and reversed
      var value = BigInteger.valueOf(100000)
      value = value.toByteArrayUnsigned().reverse();
      while (value.length < 8) value.push(0);
      for (i in testOutput.value) {
        EQ(value[i], testOutput.value[i])
      }

      // should be consistent with Transaction.addOutput()
      var test = new Transaction()
      test.addOutput('mvaRDyLUeF4CP7Lu9umbU3FxehyC5nUz3L', '100000')

      EQ(testOutput.value.length, test.outs[0].value.length)
      for (i in test.outs[0].value) {
        EQ(test.outs[0].value[i], testOutput.value[i])
      }

    })
  })
})

lib/transaction-in.js
var Script = require('btc-script');

module.exports = TransactionIn;

function TransactionIn(data) {
  // if prevout hash is null
  // then this is a coinbase transaction
  // Note: we can never have a null hash, this makes serilization of the transaction impossible
  if (!data.outpoint) {
    data.outpoint = {}
  }

  if (!data.outpoint.hash) {
    data.outpoint.hash = '0000000000000000000000000000000000000000000000000000000000000000'
    data.outpoint.index = -1
  }

  this.outpoint = data.outpoint;

  if (data.script instanceof Script) {
    this.script = data.script;
  } else {
    if (data.scriptSig) {
      this.script = Script.fromScriptSig(data.scriptSig);
    } else {
      this.script = new Script(data.script);
    }
  }
  this.sequence = data.sequence;
}

TransactionIn.prototype.clone = function() {
  var newTxin = new TransactionIn({
    outpoint: {
      hash: this.outpoint.hash,
      index: this.outpoint.index
    },
    script: this.script.clone(),
    sequence: this.sequence
  });
  return newTxin;
}


lib/transaction-out.js
var Script = require('btc-script');
var BigInteger = require('bigi');
var conv = require('binstring');

module.exports = TransactionOut;

function TransactionOut(data) {
  if (data.script instanceof Script) {
    this.script = data.script;
  } else {
    if (data.scriptPubKey) {
      this.script = Script.fromScriptSig(data.scriptPubKey);
    } else {
      this.script = new Script(data.script);
    }
  }

  if (Array.isArray(data.value)) {
    this.value = data.value;
  } else if ("string" == typeof data.value) {
    var valueHex = (new BigInteger(data.value, 10)).toString(16);
    while (valueHex.length < 16) valueHex = "0" + valueHex;
    this.value = conv(valueHex, {in: 'hex', out: 'bytes'}).reverse();
  }
};

TransactionOut.prototype.clone = function() {
  var newTxout = new TransactionOut({
    script: this.script.clone(),
    value: this.value.slice(0)
  });
  return newTxout;
};

lib/transaction.js
var Address = require('btc-address');
var BigInteger = require('bigi');
var Script = require('btc-script');
var sha256 = require('crypto-hashing').sha256;
var binConv = require('binstring');
var Parser = require('crypto-binary').MessageParser;

/*****
 TODO: 1) review https://raw2.github.com/vbuterin/bitcoinjs-lib/master/src/transaction.js
       2) import changes that make sense, signing may not make sense
******/

var _defaultNetwork = 'mainnet';
Transaction.defaultNetwork = _defaultNetwork;

Object.defineProperty(Transaction, 'defaultNetwork', {
  set: function(val) {
    _defaultNetwork = val;
    Address.defaultNetwork = val;
    Script.defaultNetwork = val;
  },
  get: function() {
    return _defaultNetwork;
  }
})

module.exports.Transaction = Transaction;

var TransactionIn = require('./transaction-in');
var TransactionOut = require('./transaction-out');
module.exports.TransactionIn = TransactionIn;
module.exports.TransactionOut = TransactionOut;

function numToVarInt(num) {
  if (num < 253) return [num];
  else if (num < 65536) return [253].concat(numToBytes(num, 2));
  else if (num < 4294967296) return [254].concat(numToBytes(num, 4));
  else return [253].concat(numToBytes(num, 8));
}

// TODO(shtylman) crypto sha uses this also
// Convert a byte array to big-endian 32-bit words
var bytesToWords = function(bytes) {
  for (var words = [], i = 0, b = 0; i < bytes.length; i++, b += 8)
    words[b >>> 5] |= bytes[i] << (24 - b % 32);
  return words;
};

// Convert big-endian 32-bit words to a byte array
var wordsToBytes = function(words) {
  for (var bytes = [], b = 0; b < words.length * 32; b += 8)
    bytes.push((words[b >>> 5] >>> (24 - b % 32)) & 0xFF);
  return bytes;
};

function numToBytes(num, bytes) {
  if (bytes === undefined) bytes = 8;
  if (bytes == 0) return [];
  else return [num % 256].concat(numToBytes(Math.floor(num / 256), bytes - 1));
}

function Transaction(doc) {
  this.version = 1;
  this.lock_time = 0;
  this.ins = [];
  this.outs = [];
  this.timestamp = null;
  this.block = null;

  if (doc) {
    if (doc.hash) this.hash = doc.hash;
    if (doc.version !== null) this.version = doc.version;
    if (doc.lock_time) this.lock_time = doc.lock_time;
    if (doc.ins && doc.ins.length) {
      for (var i = 0; i < doc.ins.length; i++) {
        this.addInput(new TransactionIn(doc.ins[i]));
      }
    }
    if (doc.outs && doc.outs.length) {
      for (var i = 0; i < doc.outs.length; i++) {
        this.addOutput(new TransactionOut(doc.outs[i]));
      }
    }
    if (doc.timestamp) this.timestamp = doc.timestamp;
    if (doc.block) this.block = doc.block;
    // if we already have the original bytes; let's keep a copy
    // see comments at deserialize()
    if (doc.serialized) this.serialized = doc.serialized
  }
};

/**
 * Turn transaction data into Transaction objects.
 *
 * Takes an array of plain JavaScript objects containing transaction data and
 * returns an array of Transaction objects.
 */
Transaction.objectify = function(txs) {
  var objs = [];
  for (var i = 0; i < txs.length; i++) {
    objs.push(new Transaction(txs[i]));
  }
  return objs;
};

/**
 * Create a new txin.
 *
 * Can be called with an existing TransactionIn object to add it to the
 * transaction. Or it can be called with a Transaction object and an integer
 * output index, in which case a new TransactionIn object pointing to the
 * referenced output will be created.
 *
 * Note that this method does not sign the created input.
 */
Transaction.prototype.addInput = function(tx, outIndex) {
  if (arguments[0] instanceof TransactionIn) {
    this.ins.push(arguments[0]);
  } else {
    this.ins.push(new TransactionIn({
      outpoint: {
        hash: tx.hash,
        index: outIndex
      },
      script: new Script(),
      sequence: 4294967295
    }));
  }
};

/**
 * Create a new txout.
 *
 * Can be called with an existing TransactionOut object to add it to the
 * transaction. Or it can be called with an Address object and a BigInteger
 * for the amount, in which case a new TransactionOut object with those
 * values will be created.
 */
Transaction.prototype.addOutput = function(address, value, network) {
  if (arguments[0] instanceof TransactionOut) {
    this.outs.push(arguments[0]);
  } else {
    if ("string" == typeof address) {
      address = new Address(address, null, network || Transaction.defaultNetwork);
    }
    if ("number" == typeof value) {
      value = BigInteger(value.toString(), 10);
    }

    if ("string" == typeof value) {
      value = BigInteger(value, 10);
    }

    if (value instanceof BigInteger) {
      value = value.toByteArrayUnsigned().reverse();
      while (value.length < 8) value.push(0);
    } else if (Array.isArray(value)) {
      // Nothing to do
    }

    this.outs.push(new TransactionOut({
      value: value,
      script: Script.createOutputScript(address, network || Transaction.defaultNetwork)
    }));
  }
};

/**
 * Serialize this transaction.
 *
 * Returns the transaction as a byte array in the standard Bitcoin binary
 * format. This method is byte-perfect, i.e. the resulting byte array can
 * be hashed to get the transaction's standard Bitcoin hash.
 */
Transaction.prototype.serialize = function() {
  var buffer = [];
  buffer = buffer.concat(wordsToBytes([parseInt(this.version)]).reverse());
  buffer = buffer.concat(numToVarInt(this.ins.length));
  for (var i = 0; i < this.ins.length; i++) {
    var txin = this.ins[i];

    buffer = buffer.concat(binConv(txin.outpoint.hash, { in : 'hex',
      out: 'bytes'
    }).reverse());
    buffer = buffer.concat(wordsToBytes([parseInt(txin.outpoint.index)]).reverse());
    var scriptBytes = txin.script.buffer;
    buffer = buffer.concat(numToVarInt(scriptBytes.length));
    buffer = buffer.concat(scriptBytes);
    buffer = buffer.concat(wordsToBytes([parseInt(txin.sequence)]).reverse());
  }
  buffer = buffer.concat(numToVarInt(this.outs.length));
  for (var i = 0; i < this.outs.length; i++) {
    var txout = this.outs[i];
    buffer = buffer.concat(txout.value);
    var scriptBytes = txout.script.buffer;
    buffer = buffer.concat(numToVarInt(scriptBytes.length));
    buffer = buffer.concat(scriptBytes);
  }
  buffer = buffer.concat(wordsToBytes([parseInt(this.lock_time)]).reverse());

  return buffer;
};

var OP_CODESEPARATOR = 171;

var SIGHASH_ALL = 1;
var SIGHASH_NONE = 2;
var SIGHASH_SINGLE = 3;
var SIGHASH_ANYONECANPAY = 80;

/**
 * Hash transaction for signing a specific input.
 *
 * Bitcoin uses a different hash for each signed transaction input. This
 * method copies the transaction, makes the necessary changes based on the
 * hashType, serializes and finally hashes the result. This hash can then be
 * used to sign the transaction input in question.
 */
Transaction.prototype.hashTransactionForSignature = function(connectedScript, inIndex, hashType) {
  var txTmp = this.clone();

  // In case concatenating two scripts ends up with two codeseparators,
  // or an extra one at the end, this prevents all those possible
  // incompatibilities.
  /*scriptCode = scriptCode.filter(function (val) {
 return val !== OP_CODESEPARATOR;
 });*/

  // Blank out other inputs' signatures
  for (var i = 0; i < txTmp.ins.length; i++) {
    txTmp.ins[i].script = new Script();
  }

  txTmp.ins[inIndex].script = connectedScript;

  // Blank out some of the outputs
  if ((hashType & 0x1f) == SIGHASH_NONE) {
    txTmp.outs = [];

    // Let the others update at will
    for (var i = 0; i < txTmp.ins.length; i++)
      if (i != inIndex)
        txTmp.ins[i].sequence = 0;
  } else if ((hashType & 0x1f) == SIGHASH_SINGLE) {
    // TODO: Implement
  }

  // Blank out other inputs completely, not recommended for open transactions
  if (hashType & SIGHASH_ANYONECANPAY) {
    txTmp.ins = [txTmp.ins[inIndex]];
  }

  var buffer = txTmp.serialize();

  //todo: change to buffer
  buffer = buffer.concat(wordsToBytes([parseInt(hashType)]).reverse());
  return sha256.x2(buffer, { in : 'bytes',
    out: 'bytes'
  });
}

/**
 * Calculate and return the transaction's hash.
 */
Transaction.prototype.getHash = function() {
  // todo: migrate to buffer
  if (Array.isArray(this.serialized) && this.serialized.length > 0) {
    var buffer = this.serialized;
  } else {
    var buffer = this.serialize();
  }

  //todo: change to buffer
  return sha256.x2(buffer, { in : 'bytes',
    out: 'bytes'
  }).reverse();
};

Transaction.prototype.getSize = function() {
  // todo: migrate to buffer
  if (Array.isArray(this.serialized) && this.serialized.length > 0) {
    var buffer = this.serialized;
  } else {
    var buffer = this.serialize();
  }

  //todo: change to buffer
  return buffer.length;
}

/**
 * Create a copy of this transaction object.
 */
Transaction.prototype.clone = function() {
  var newTx = new Transaction();
  newTx.version = this.version;
  newTx.lock_time = this.lock_time;
  for (var i = 0; i < this.ins.length; i++) {
    var txin = this.ins[i].clone();
    newTx.addInput(txin);
  }
  for (var i = 0; i < this.outs.length; i++) {
    var txout = this.outs[i].clone();
    newTx.addOutput(txout);
  }
  return newTx;
};

/**
 * Analyze how this transaction affects a wallet.
 *
 * Returns an object with properties 'impact', 'type' and 'addr'.
 *
 * 'impact' is an object, see Transaction#calcImpact.
 *
 * 'type' can be one of the following:
 *
 * recv:
 *   This is an incoming transaction, the wallet received money.
 *   'addr' contains the first address in the wallet that receives money
 *   from this transaction.
 *
 * self:
 *   This is an internal transaction, money was sent within the wallet.
 *   'addr' is undefined.
 *
 * sent:
 *   This is an outgoing transaction, money was sent out from the wallet.
 *   'addr' contains the first external address, i.e. the recipient.
 *
 * other:
 *   This method was unable to detect what the transaction does. Either it
 */
/*Transaction.prototype.analyze = function(wallet) {
  var Wallet = require('./wallet');
  if (!(wallet instanceof Wallet)) return null;

  var allFromMe = true,
    allToMe = true,
    firstRecvHash = null,
    firstMeRecvHash = null,
    firstSendHash = null;

  for (var i = this.outs.length - 1; i >= 0; i--) {
    var txout = this.outs[i];
    var hash = txout.script.simpleOutPubKeyHash();
    if (!wallet.hasHash(hash)) {
      allToMe = false;
    } else {
      firstMeRecvHash = hash;
    }
    firstRecvHash = hash;
  }
  for (var i = this.ins.length - 1; i >= 0; i--) {
    var txin = this.ins[i];
    firstSendHash = txin.script.simpleInPubKeyHash();
    if (!wallet.hasHash(firstSendHash)) {
      allFromMe = false;
      break;
    }
  }

  var impact = this.calcImpact(wallet);

  var analysis = {};

  analysis.impact = impact;

  if (impact.sign > 0 && impact.value.compareTo(BigInteger.ZERO) > 0) {
    analysis.type = 'recv';
    analysis.addr = new Bitcoin.Address(firstMeRecvHash);
  } else if (allFromMe && allToMe) {
    analysis.type = 'self';
  } else if (allFromMe) {
    analysis.type = 'sent';
    // TODO: Right now, firstRecvHash is the first output, which - if the
    //       transaction was not generated by this library could be the
    //       change address.
    analysis.addr = new Bitcoin.Address(firstRecvHash);
  } else {
    analysis.type = "other";
  }

  return analysis;
};*/

/**
 * Get a human-readable version of the data returned by Transaction#analyze.
 *
 * This is merely a convenience function. Clients should consider implementing
 * this themselves based on their UI, I18N, etc.
 */
/*Transaction.prototype.getDescription = function(wallet) {
  var analysis = this.analyze(wallet);

  if (!analysis) return "";

  switch (analysis.type) {
    case 'recv':
      return "Received with " + analysis.addr;
      break;

    case 'sent':
      return "Payment to " + analysis.addr;
      break;

    case 'self':
      return "Payment to yourself";
      break;

    case 'other':
    default:
      return "";
  }
};*/

/**
 * Get the total amount of a transaction's outputs.
 */
Transaction.prototype.getTotalOutValue = function() {
  var totalValue = BigInteger.ZERO;
  for (var j = 0; j < this.outs.length; j++) {
    var txout = this.outs[j];
    totalValue = totalValue.add(BigInteger.fromByteArrayUnsigned(txout.value.reverse()));
  }
  return totalValue;
};

/**
 * Old name for Transaction#getTotalOutValue.
 *
 * @deprecated
 */
Transaction.prototype.getTotalValue = Transaction.prototype.getTotalOutValue;

// TODO: Make pull request in cypto-binary
Parser.prototype.readVarRaw = function readVarRaw() {
  if (this.hasFailed || this.pointerCheck() === false) return false;
  var flagRaw = this.raw(1);
  if (flagRaw) {
    var flag = flagRaw.readUInt8(0)
  } else {
    return false
  }

  if (flag < 0xfd) {
    return flagRaw;
  } else
  if (flag == 0xfd) {
    return Buffer.concat([flagRaw, this.raw(2)]);
  } else if (flag == 0xfe) {
    return Buffer.concat([flagRaw, this.raw(4)]);
  } else {
    return Buffer.concat([flagRaw, this.raw(8)]);
  }
};

// Second argument txCount is optional; It indicates
// the number of the transactions that you expect to
// parse. The default number is 1
Transaction.deserialize = function(buf, txCount) {
  if (!Buffer.isBuffer(buf)) {
    buf = binConv(buf, {
      out: 'buffer'
    });
  }

  var s = new Parser(buf)

  if (!txCount) {
    txCount = 1;
  }

  var transactions = []
  for (var y = 0; y < txCount; y++) {
    var doc = {};

    // note: serializing a large transaction (1000+ inputs) such as testnet tx ('95ea61f319ed0d2b28e94cb0164396b4024bc6ad624fcb492c5c87a088592e81') can take up to 2 minutes. So we have to cache the original bytes when we deserialize it so that getHash() will be fast

    // this will be concatenated into a buffer in the end
    doc.serialized = []
    var verB = s.raw(4)
    doc.serialized.push(verB)
    doc.version = new Parser(verB).readUInt32LE()

    doc.ins = []
    doc.outs = []

    var inB = s.readVarRaw()
    doc.serialized.push(inB)
    var inputCount = new Parser(inB).readVarInt()
    for (var i = 0; i < inputCount; i++) {
      var data = {}
      data.outpoint = {}

      var txHash = s.raw(32)
      if (txHash === false) {
        return false
      }
      doc.serialized.push(txHash)

      // TODO: change this to buffer in the transition to buffer
      data.outpoint.hash = binConv(binConv(txHash, {
        out: 'bytes'
      }).reverse(), {
        out: 'hex'
      })

      isCoinbase = (data.outpoint.hash === '0000000000000000000000000000000000000000000000000000000000000000')

      var indexB = s.raw(4)
      doc.serialized.push(indexB)
      data.outpoint.index = new Parser(indexB).readUInt32LE()
      // 0xFFFFFFFF (4294967295) is -1 when trying to read it as UInt32
      if (data.outpoint.index === 4294967295) {
        data.outpoint.index = -1
      }

      var sLenB = s.readVarRaw()
      doc.serialized.push(sLenB)
      var scriptLength = new Parser(sLenB).readVarInt()

      var sB = s.raw(scriptLength)
      doc.serialized.push(sB)
      data.script = new Script(sB.toString('hex'), isCoinbase)

      var seqB = s.raw(4)
      doc.serialized.push(seqB)
      data.sequence = new Parser(seqB).readUInt32LE()

      doc.ins.push(data)
    }

    var outB = s.readVarRaw()
    doc.serialized.push(outB)
    var outputCount = new Parser(outB).readVarInt()
    for (var i = 0; i < outputCount; i++) {
      var data = {}
      var value = s.raw(8)
      if (value === false) {
        return false
      }
      doc.serialized.push(value)

      data.value = binConv(value, {
        out: 'bytes'
      })

      var sLenB = s.readVarRaw()
      doc.serialized.push(sLenB)
      var scriptLength = new Parser(sLenB).readVarInt()

      var sB = s.raw(scriptLength)
      doc.serialized.push(sB)
      data.script = new Script(sB.toString('hex'))

      doc.outs.push(data)
    }

    var lockB = s.raw(4)
    doc.serialized.push(lockB)
    doc.lock_time = new Parser(lockB).readUInt32LE()

    // now we turn serialized into a buffer
    doc.serialized = binConv(Buffer.concat(doc.serialized), { in : 'buffer',
      out: 'bytes'
    })

    if (s.hasFailed) {
      return false
    } else {
      transactions.push(new Transaction(doc))
    }
  }

  if (s.hasFailed) {
    return false
  }

  if (txCount === 1) {
    return transactions[0]
  } else {
    return transactions
  }
}